fluent-syntax/src/parser/ftlstream.rs

```diff
@@ -1,394 +1,211 @@
-        while let Some(_) = self.next() {
-            if self.current_is('\n') && !self.peek_char_is('\n') {
-                self.next();
-
-                if self.ch.is_none()
-                    || self.is_entry_id_start()
-                    || self.current_is('#')
-                    || (self.current_is('/') && self.peek_char_is('/'))
-                    || (self.current_is('[') && self.peek_char_is('['))
-                {
-                    break;
-                }
-            }
+    pub fn is_identifier_start(&self) -> bool {
+        match self.source.get(self.ptr) {
+            Some(b) if self.is_byte_alphabetic(*b) => true,
+            _ => false,
         }
     }
 
-    fn take_id_start(&mut self, allow_term: bool) -> Result<char> {
-        if allow_term && self.current_is('-') {
-            self.next();
-            return Ok('-');
-        }
+    pub fn is_byte_alphabetic(&self, b: u8) -> bool {
+        (b >= b'a' && b <= b'z') || (b >= b'A' && b <= b'Z')
+    }
 
-        if let Some(ch) = self.ch {
-            if self.is_char_id_start(Some(ch)) {
-                let ret = self.ch.unwrap();
-                self.next();
-                return Ok(ret);
-            }
-        }
+    pub fn is_byte_digit(&self, b: u8) -> bool {
+        b >= b'0' && b <= b'9'
+    }
 
-        let allowed_range = if allow_term {
-            "'a'...'z' | 'A'...'Z' | '-'"
-        } else {
-            "'a'...'z' | 'A'...'Z'"
-        };
-        error!(ErrorKind::ExpectedCharRange {
-            range: String::from(allowed_range),
-        })
+    pub fn is_byte_pattern_continuation(&self, b: u8) -> bool {
+        ![b'}', b'.', b'[', b'*'].contains(&b)
     }
 
-    fn take_id_char(&mut self) -> Option<char> {
-        let closure = |x| match x {
-            'a'...'z' | 'A'...'Z' | '0'...'9' | '_' | '-' => true,
+    pub fn is_number_start(&self) -> bool {
+        match self.source.get(self.ptr) {
+            Some(b) if (b == &b'-') || self.is_byte_digit(*b) => true,
             _ => false,
-        };
-
-        match self.take_char(closure) {
-            Some(ch) => Some(ch),
-            None => None,
         }
     }
 
-    fn take_variant_name_char(&mut self) -> Option<char> {
-        let closure = |x| match x {
-            'a'...'z' | 'A'...'Z' | '0'...'9' | '_' | '-' | ' ' => true,
+    pub fn is_eol(&self) -> bool {
+        match self.source.get(self.ptr) {
+            Some(b'\n') => true,
+            Some(b'\r') if self.is_byte_at(b'\n', self.ptr + 1) => true,
             _ => false,
-        };
-
-        match self.take_char(closure) {
-            Some(ch) => Some(ch),
-            None => None,
         }
     }
 
-    fn take_digit(&mut self) -> Option<char> {
-        let closure = |x| match x {
-            '0'...'9' => true,
-            _ => false,
-        };
+    pub fn get_slice(&self, start: usize, end: usize) -> &'p str {
+        unsafe { str::from_utf8_unchecked(&self.source[start..end]) }
```

Comment:
The consequence is that I'm using our collective brain power, instead of compiler's brain power, to ensure that I'm not slicing outside of the boundary.
The boundary checks are relatively expensive and since we're a full byte parser I believe we always attempt to slice within the range.

So, it's not about whether the input is valid UTF8, it's only about whether the parser always selects the right start/end combo that's within the source slice.

I'm going to use a fuzzer to further verify that claim, but I'd prefer not to overblow the lib with options that don't really mean too much to the user. Currently if you pick wrong start/end, the parser crashes, but if we instead use `from_utf8` we'll likely use it with `expect` which will panic as well :)

If while working on 1.0 we'll discover that we have a lot of places in the parser that pick start/end wrong we can try to mitigate it here, but I'd honestly prefer to fix them in the parser.

Please, not that unsafe code is not "bad" per design. It makes total sense to use it in many cases, and people like Henri are actively asking for improvements in how we calculate boundary check requirements to minimize when we need them: https://hsivonen.fi/rust2019/ .

Our problem is that we have code like:

```rust
let source = "id = Foo".as_bytes();
let ptr = 0;
while let Some(b) = source.get(ptr) {
    if b == b'i' {
        if let Some(b2) = source.get(ptr + 1) {
            if b2 == b'd' {
              let id = unsafe { str::from_utf8_unchecked(ptr, ptr+1); }
            }
        }
    }
    ptr += 1;
}
```

(sorry for how stupid the code is). It would be insanely hard for the compiler to deduct that it can use boundary check elision, but we can deduct that quite easily - we just checked for what is at `ptr`, and what is at `ptr+1`.

Reply:
why not store the string as `&str` and use `as_bytes()` in your processing code everywhere, so that you can safely subslice it with byte indices?

Reply:
Digging a bit deeper into this, the panic seems to originate from `&self.source[start..end]` rather than the unsafe call to `from_utf8_unchecked`. Is there a checked way to index into a slice?

Reply:
OK, so I think there are two things going on here. Can you please confirm that my understanding is correct?

1. When creating the slice of bytes with `&self.source[start..end]`, `end` might go out of bounds. We should check that it isn't or maybe use an iterator and the checked `slice::get` method. The `x = {"\\u` testcase exposes this bug.

2. This new slice is then passed to `from_utf8_unchecked` which is an unsafe method. If the slice of bytes contains a sequence of bytes which is invalid in UTF8, this method will panic. Realistically this is unlikely to happen with files which are read from disk and validated as UTF8, but might still be an issue when the data is not guaranteed to be UTF8.

Reply:
> (…) might still be an issue when the data is not guaranteed to be UTF8.

Is this even possible given that ` FluentResource::try_new` takes a `String`?

Reply:
> The x = {"\\u testcase exposes this bug.

I fixed it and reviewed other places where this may happen. Found no other so I think we're good.

> This new slice is then passed to from_utf8_unchecked which is an unsafe method. If the slice of bytes contains a sequence of bytes which is invalid in UTF8, this method will panic

No, we don't care if the file is UTF8 or if it contains invalid UTF8 tokens because the boundaries we set are validated to be. In other words, since Fluent tokens end on valid UTF8 chars like "\n", "{" etc. we can establish a range and know that we're slicing at the right place. What's inside, doesn't really matter to us. So, if the entry looks like this:

```
key = This is a XXX value { $foo }
```

then we slice out three times:
1) `key` which has a `start` and `end` known to be a valid alpha character and every character inside as well.
2) `This is a XXX value` which has `start` as the first non-space char, and `end` as one-before `{` char, so also no risk of cutting in the middle of a multi-byte char.
3) `foo` - similar to (1)

Even if in (2) the XXX is invalid UTF8, we don't care about it. It still slices properly because of what is a boundary in Fluent.